### PR TITLE
[Bug fix] Make routed-expert cache completeness dtype-aware

### LINKED ISSUE #54515: [model]: routed-expert cache completeness ignores expert dtype
### Component / Area

model — `models/demos/deepseek_v3_d_p` (prefill, MoE expert cache)

### Issue Type

Bad Outputs

**Found during:** Mistral Small 4 119B prefill bring-up (#53688). The bug is in shared DeepSeek-family prefill code — any variant whose expert dtype differs from a cache on disk hits it, not just Mistral.

## Observed

`as_tensor` stamps the requested dtype into the `.tensorbin` filename, but the completeness check globs **without** it. A cache built at `bfloat4_b` therefore answers "complete" for a `bfloat8_b` request, regeneration is skipped, and the empty placeholder is loaded **as the weights** — plausible output, meaningless PCC, no error anywhere.

The default expert dtype was also spelled as a literal in six separate places (`TtMoe`, `TtPrefillBlock` ×2, `TtPrefillTransformer`, `TtPrefillRuntime`, and the layer-by-layer loader), so build and check could disagree without anyone selecting a non-default dtype.

## Expected

A cache written at a different expert dtype is reported missing and rebuilt.

## Steps to Reproduce

1. Build a routed-expert TTNN cache for a variant at one expert dtype.
2. Run any prefill accuracy test for a variant whose expert dtype differs, pointing `TT_MISTRAL4_PREFILL_TTNN_CACHE` (or the DeepSeek equivalent) at that cache.
3. The run reports the cache complete, loads placeholders, and reports a PCC that does not reflect the weights.

**Input data:** any routed-expert cache on disk plus a variant requesting a different expert dtype.
**Frequency:** Always, whenever the on-disk cache dtype differs from the requested one.

## Software Versions

- tt-metal main `58342a6f63d` (2026-08-26) — bug present at this commit
- `transformers == 5.12.1`, `torch == 2.11.0+cpu` (repo pins)
- Ubuntu 22.04.5 LTS

## Hardware Details

- Product: Blackhole
- Card/System: Galaxy, 32-chip (`bh-glx-110-a04u02`)

## Regression

Unknown — present since the routed-expert cache gained a dtype-qualified filename while the completeness check did not.

## Priority

P2

## Impact

Any variant whose expert dtype differs from a cache already on disk. Silently invalidates accuracy results, which is worse than a crash: the number looks plausible.

### PR DESCRIPTION
**Pull request by kmabeeTT**
### PR Category
Bug fix

### Summary

Fixes #54515.

Thread the requested expert dtype through the cache-completeness path and match on the dtype-qualified filename, so a cache written at a different dtype is reported missing and rebuilt instead of being loaded as placeholders.

Pin the dtype the check will later request, threaded runtime → transformer → block → MoE, with the default in `utils/expert_dtypes.py` so build and check cannot disagree. It was previously a literal in six places.

Routed experts only — the other cached modules still glob dtype-blind, which is a latent instance of the same bug and deliberately out of scope here.

Split out of #53747 per review; one fix per PR.

### Notes for reviewers

- The behavioural change is in `tt_routed_expert.py` / `tt_moe.py`: completeness is now answered for a *specific* dtype rather than for "any file with this prefix".
- Most of the diff is threading one parameter through four layers; the load-bearing part is the filename match.
- `tests/cache/test_routed_expert_cache.py` covers both directions — matching dtype reported complete, differing dtype reported missing.
- Rebased onto `58342a6f63d`; no conflict with the SiTU changes in `tt_moe.py` from #54149.

### Testing

Local, on a 32-chip Blackhole galaxy, on a branch carrying all five split fixes on this PR's base:

- `tests/cache/test_routed_expert_cache.py` — matching dtype reported complete, differing dtype reported missing
- Mistral Small 4 MoE accuracy + perf, 8x4, 5k and 25k — 4 passed (`final_output` 0.986391 vs 0.982 bar, `reference_output` 0.994691, `shared_output` 0.999761, `routed_output` 0.937314)
- Mistral Small 4 prefill block, 8x4 — 6 passed (caches built and reloaded through the changed path)
- Blaze dispatched on this branch, `test-type=glm_moe,kimi_moe,kimi_k3_moe,prefill_block,prefill_runner,prefill_runner_kv_pcc`: [run 33043217007](https://github.com/tenstorrent/tt-metal/actions/runs/33043217007) (15/16 — the GLM-5.2 disaggregated producer KV leg reported PCC -0.00081 once; **a flake**: [re-run of that leg on the identical commit passes](https://github.com/tenstorrent/tt-metal/actions/runs/33079416117), and a control on this PR's base commit `58342a6f63d` also passes ([run 33079707414](https://github.com/tenstorrent/tt-metal/actions/runs/33079707414)). The abort guard added here never fired in the failing run, and the cache reported complete on all ranks.)
- These jobs were green on `3900b5ea463`, one commit below this base. The overall run there is red only from `high_bw_all_gather CCL perf + accuracy`, a pre-existing flake outside this selection.
- All five together on `kmabee/issue_54515-54519_integration` (base `58342a6f63d`): Blaze [`test-type=all`](https://github.com/tenstorrent/tt-metal/actions/runs/33094428995) green 44/44, [(Blackhole) e2e](https://github.com/tenstorrent/tt-metal/actions/runs/33094442486) 39/40.
- The one red is not from these changes: the DeepSeek perf threshold has 0.15% headroom against 0.41% machine variance ([clean-base control on the same machine passed](https://github.com/tenstorrent/tt-metal/actions/runs/33115646268)).

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:kmabee/issue_54515_fix-moe-expert-cache-dtype)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:kmabee/issue_54515_fix-moe-expert-cache-dtype)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:kmabee/issue_54515_fix-moe-expert-cache-dtype)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:kmabee/issue_54515_fix-moe-expert-cache-dtype)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:kmabee/issue_54515_fix-moe-expert-cache-dtype)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:kmabee/issue_54515_fix-moe-expert-cache-dtype)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:kmabee/issue_54515_fix-moe-expert-cache-dtype)

### CODE REVIEW

File: models/demos/deepseek_v3_d_p/tt/tt_prefill_runtime.py [hunk 1]
```diff
@@ -184,6 +185,9 @@ def _build_model(self, state_dict: dict) -> None:
                 # to look for the latent-projection cache files and would call an incomplete cache
                 # complete. model_cfg is already in hand two lines up.
                 model_cfg=model_cfg,
+                # Must be the dtype the experts will be BUILT at, or a cache at another dtype
+                # reports complete and the placeholder is loaded as the weights.
+                routed_expert_weights_dtype=self.config.routed_expert_weights_dtype,
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply models/demos/deepseek_v3_d_p/tt/tt_prefill_runtime.py:190]: Good catch, and correct: the warning claimed a failure that never happened. Fixed in cb343d16a28 — it now raises when the cache is incomplete *and* no source weights were supplied. Runs that do provide weights still rebuild the cache as before. `prefill_runner` / `prefill_runner_kv_pcc` added to this PR's CI selection since they exercise the empty-state_dict path.

File: models/demos/deepseek_v3_d_p/utils/expert_dtypes.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+"""The default routed-expert weight dtype, in one place.
+
+`routed_expert_weights_dtype=ttnn.bfloat4_b` used to be spelled as a literal default in six places
+(TtMoe, TtPrefillBlock x2, TtPrefillTransformer, TtPrefillRuntime, and the layer-by-layer loader).
+A/B-ing expert precision therefore meant editing six defaults and hoping none was missed -- and
+missing one is not a crash, it is a weight cache BUILT at one dtype and CHECKED at another, which
+loads the empty placeholder as the weights and produces plausible output with a meaningless PCC.
+
+One constant, so the build and the completeness check cannot disagree. Callers that want a different
+precision pass `routed_expert_weights_dtype` explicitly.
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply models/demos/deepseek_v3_d_p/utils/expert_dtypes.py]: Partly addressed. The docstring overclaimed and is corrected in 27f18c88099 to say plainly that `load_and_compute_layer_by_layer` is *not* migrated and still carries a literal. Not closing it here: that loader lives in `transformer_helpers.py`, which is owned by #54528 and #54529 in this split, so migrating it from this PR would put one file in three open PRs. Leaving this thread open for the reviewer to weigh in on scope.

**Comment by kmabeeTT** [comment]
Testing section updated with CI links. Targeted Blaze run on this branch is green, and all five split fixes were also raised together on a combined branch (`kmabee/issue_54515-54519_integration`, same base) to check they compose: Blaze `test-type=all` green 44/44, plus (Blackhole) e2e. The one e2e red there is a perf threshold with 0.15% headroom against 0.41% machine variance — a clean-base control on the same machine passed, so it is not from these changes.

File: models/demos/deepseek_v3_d_p/utils/expert_dtypes.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+NOT yet migrated: `load_and_compute_layer_by_layer` in utils/transformer_helpers.py still carries a
+literal `ttnn.bfloat4_b` default, and none of its four call sites override it. Changing the constant
+below therefore leaves that loader building at the old dtype -- the very mismatch this module
+exists to prevent. It is left for a follow-up because transformer_helpers.py belongs to a different
+PR in this split; migrating it here would put the same file in three PRs at once.
+"""
+
+import ttnn
+
+DEFAULT_ROUTED_EXPERT_WEIGHTS_DTYPE = ttnn.bfloat4_b
```
Comment (pmilojevicTT) [inline models/demos/deepseek_v3_d_p/utils/expert_dtypes.py]: Can we just put this constant inside `tt_routed_expert.py` file instead of making new file, just for one new constant, thanks.

File: models/demos/deepseek_v3_d_p/tt/moe/tt_routed_expert.py [hunk 4]
```diff
@@ -49,15 +50,33 @@
...
+        """
         from models.demos.deepseek_v3_d_p.utils.fast_cache_checker import pattern_exists
 
+        dtype_name = weights_dtype.name
         for local_expert_idx in range(experts_per_chip):
             for proj in ["gate", "up", "down"]:
-                pattern = f"{cache_name_prefix}.local_{local_expert_idx}_{proj}*.tensorbin"
-                if not pattern_exists(pattern, "RoutedExpert"):
-                    logger.debug(f"TTNN cache missing: {cache_name_prefix}.local_{local_expert_idx}_{proj}")
+                stem = f"{cache_name_prefix}.local_{local_expert_idx}_{proj}"
```
Comment (pmilojevicTT) [inline models/demos/deepseek_v3_d_p/tt/moe/tt_routed_expert.py:88]: @mbezuljTT @kgrujcic-tt Can this adding new prefix for RE weight cache potentially make problem if we already have stored routed expert weights for various models with old cache prefix? Will this implementation just skip those premade weights inside `/data` exabox folder and start making new unnecessary weights, so we will have double weights on NFS disk?

Comment (kgrujcic-tt) [reply models/demos/deepseek_v3_d_p/tt/moe/tt_routed_expert.py:88]: > The default matches the routed-expert default (bfloat4_b) so existing callers are unaffected; any caller that builds at a different dtype must pass the same value here.

I didn't observe anything unusual on `/data`, so I'd say it's fine.

**Review by kgrujcic-tt — APPROVED** [review]

File: models/demos/deepseek_v3_d_p/utils/expert_dtypes.py (hunk 3 above)
Comment (kmabeeTT) [reply models/demos/deepseek_v3_d_p/utils/expert_dtypes.py]: Done — moved to tt_routed_expert.py and expert_dtypes.py is gone.

File: models/demos/deepseek_v3_d_p/tt/moe/tt_routed_expert.py (hunk 4 above)
Comment (kmabeeTT) [reply models/demos/deepseek_v3_d_p/tt/moe/tt_routed_expert.py:88]: FWIW  this PR doesn't change any filename, only what the completeness check globs for... as_tensor has always encoded the dtype, so the files on disk already look like `layer_6.routed_expert.local_0_gate_dtype_BFLOAT4_B_layout_TILE.tensorbin` — the old glob was just dtype-blind and matched any of them.

Checked against a real cache dir: the new pattern matches 162/162 existing local_0_gate files, so nothing is skipped or rebuilt.

**Review by pmilojevicTT — APPROVED** [review]
